### PR TITLE
[Ncl-1983] Allow SCMImpl to clone different repositories in parallel

### PR DESCRIPTION
**Pull request by stanlyDoge**

### CODE REVIEW

**Comment by stanlyDoge** [comment]
retest this please

File: source-code-manager/src/main/java/org/jboss/da/scm/impl/DirectoryReference.java [hunk 1]
```diff
@@ -5,17 +5,21 @@
...
  *
  * @author Honza Brázdil <jbrazdil@redhat.com>
  */
 class DirectoryReference {
 
-    Logger log = LoggerFactory.getLogger(DirectoryReference.class);
+    @Inject
+    Logger log;
+
+    //Logger log = LoggerFactory.getLogger(DirectoryReference.class);
```
Comment (janinko) [inline source-code-manager/src/main/java/org/jboss/da/scm/impl/DirectoryReference.java]: Here must stay the LoggerFactory, becuse DirectoryReference is not a bean (the container won't inject anything into it).

File: source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMCache.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                } catch (InterruptedException | ExecutionException | TimeoutException ex) {
+                    log.warn("InterruptedException | ExecutionException | TimeoutException");
+                }
+            }
+        }
+    }
+
+    @Schedule(hour = "*", minute = "*/10")
+    public void invalidateCache() {
+        Iterator<Map.Entry<SCMSpecifier, DirectoryReference>> it = cache.entrySet().iterator();
```
Comment (janinko) [inline source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMCache.java]: Call checkFutureCache first

File: source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMCache.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    private void checkFutureCache() {
+        Iterator<Map.Entry<SCMSpecifier, FutureReference>> it = futureCache.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<SCMSpecifier, FutureReference> e = it.next();
+            FutureReference value = e.getValue();
+            if (value.isDone()) {
+                try {
+                    cache.put(e.getKey(), value.get(1, TimeUnit.SECONDS));
+                } catch (InterruptedException | ExecutionException | TimeoutException ex) {
+                    log.warn("InterruptedException | ExecutionException | TimeoutException");
```
Comment (janinko) [inline source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMCache.java]: There could be better log message. Also pass the exception to the logger.

File: source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMCache.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    private void checkFutureCache() {
+        Iterator<Map.Entry<SCMSpecifier, FutureReference>> it = futureCache.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<SCMSpecifier, FutureReference> e = it.next();
+            FutureReference value = e.getValue();
+            if (value.isDone()) {
+                try {
+                    cache.put(e.getKey(), value.get(1, TimeUnit.SECONDS));
```
Comment (janinko) [inline source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMCache.java]: when you move the reference to the other cache, you should remove it from the firts.

File: source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMImpl.java [hunk 5]
```diff
@@ -30,59 +27,45 @@
...
-        Iterator<Map.Entry<SCMSpecifier, DirectoryReference>> it = cache.entrySet().iterator();
-        while (it.hasNext()) {
-            DirectoryReference ref = it.next().getValue();
-            if (!ref.check()) {
-                it.remove();
+        } else {
+            try {
+                fref.get(30, TimeUnit.MINUTES);
+            } catch (InterruptedException | ExecutionException | TimeoutException ex) {
+                log.warn("InterruptedException | ExecutionException | TimeoutException");
```
Comment (janinko) [inline source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMImpl.java]: Use better error message.

File: source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMImpl.java [hunk 6]
```diff
@@ -30,59 +27,45 @@
...
-    }
-
-    @PreDestroy
-    private void cleanup() {
-        Iterator<Map.Entry<SCMSpecifier, DirectoryReference>> it = cache.entrySet().iterator();
-        while (it.hasNext()) {
-            it.next().getValue().delete();
-            it.remove();
-        }
+        return null;
```
Comment (janinko) [inline source-code-manager/src/main/java/org/jboss/da/scm/impl/SCMImpl.java]: Why do you return null here?

**Comment by janinko** [comment]
Please squash the two commits together.

**Comment by stanlyDoge** [comment]
closed due to pull request #334

**Pull request closed without merging**